### PR TITLE
[CONTINT-3853][Dogstatsd] Add support of origin detection with container name + pod uid

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR adds origin detection support with the pod uid and the container name for dogstatsd.

### Describe how to test/QA your changes
https://dddev.datadoghq.com/notebook/7840405/origin-detection-dogstatsd?range=900000&start=1712661471000&live=true

Deploy the agent, when the agent is ready, deploy the following app:
```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dogstatsd-udp
  labels:
    app: dogstatsd-udp
spec:
  replicas: 1
  selector:
    matchLabels:
      app: dogstatsd-udp
  template:
    metadata:
      labels:
        app: dogstatsd-udp
    spec:
      initContainers:
        - name: app
          image: ubuntu
          command:
            - /bin/bash
            - -c
            - |
              export DEBIAN_FRONTEND=noninteractive
              apt update
              apt install -y netcat bc

              for i in $(seq 1 10); do
                sleep 5
                printf 'initcustom.metric.name:%s|g|#dd.internal.entity_id:%s,foo:bar\n' "$(echo "s($(date +%s)/60)" | bc -l)" "$DD_ENTITY_ID" | nc -u -q 0 -w60 "$DD_AGENT_HOST" 8125
              done
          env:
            - name: DD_AGENT_HOST
              valueFrom:
                fieldRef:
                  fieldPath: status.hostIP
            - name: DD_INTERNAL_POD_UID
              valueFrom:
                fieldRef:
                  fieldPath: metadata.uid
            - name: DD_ENTITY_ID
              value: "en-init.$(DD_INTERNAL_POD_UID)/app"
      containers:
        - name: appp
          image: ubuntu
          command:
            - /bin/bash
            - -c
            - |
              export DEBIAN_FRONTEND=noninteractive
              apt update
              apt install -y netcat bc

              while true; do
                while true; do
                  printf 'custom.metric.name:%s|g|#dd.internal.entity_id:%s,foo:bar\n' "$(echo "s($(date +%s)/60)" | bc -l)" "$DD_ENTITY_ID" 
                  sleep 10
                done | nc -u -q0 -w60 "$DD_AGENT_HOST" 8125
              done
          env:
            - name: DD_AGENT_HOST
              valueFrom:
                fieldRef:
                  fieldPath: status.hostIP
            - name: DD_INTERNAL_POD_UID
              valueFrom:
                fieldRef:
                  fieldPath: metadata.uid
            - name: DD_ENTITY_ID
              value: "en-$(DD_INTERNAL_POD_UID)/appp"

```
Make sure the metrics are tagged with container tags.

### CODE REVIEW

File: comp/core/tagger/tagger.go [hunk 1]
```diff
@@ -442,6 +446,42 @@ func (t *TaggerClient) EnrichTags(tb tagset.TagsAccumulator, originInfo taggerty
...
+	case 4:
+		podUID = parts[2]
+		cname = parts[3]
+		initCont = parts[1] == "init."
+	}
+	cid, err := metricsProvider.ContainerIDForPodUIDAndContName(
+		podUID,
+		cname,
+		initCont,
+		time.Second,
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply comp/core/tagger/tagger.go:476]: I just wrote `timeout` by mistake. I am not sure about what would be the best cache validity. 

I would assume that the shorter the better. Indeed, `ContainerIDForPodUIDAndContName` only retrieves the pod from workloadmeta and returns the container-id matching the container name so the cost of the operation depends on workloadmeta's performances but should be reasonnable. 

On the one hand, if the cache validity is long, it means that if the container-id changes, `ContainerIDForPodUIDAndContName` will return an outdated container-id and the metrics will be wrongly tagged.
On the other hand, if it's too short then we might lose all tags until the workloadmeta collector polls the kubelet and updates the pod entry again.

File: comp/core/tagger/tagger.go [hunk 2]
```diff
@@ -442,6 +446,42 @@ func (t *TaggerClient) EnrichTags(tb tagset.TagsAccumulator, originInfo taggerty
 
 }
 
+type cidRetriever interface {
+	// ContainerIDForPodUIDAndContName returns the container ID for a given pod UID and container name
+	ContainerIDForPodUIDAndContName(podUID, contName string, isInit bool, timeout time.Duration) (string, error)
+}
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply comp/core/tagger/tagger.go]: Indeed, that's a good catch. I removed it then.

**Review by L3n41c — APPROVED** [review]

**Pull request merged**